### PR TITLE
lint: replace `isort` with Ruff's rule I

### PR DESCRIPTION
**Pull request by Borda**
## Description

since #15312 moved formatting from Black to Rufft, it would make sense also unify import formatting under the same ruff so use build-in `I` rule instead of additional `isort`

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by bdice** [comment]
Hi @Borda, thanks for this. However, we have been keeping isort because ruff does not support Cython code but isort does. See this discussion for more info: https://github.com/rapidsai/cudf/issues/14882#issuecomment-1910887340

**Comment by vyasr** [comment]
> Hi @Borda, thanks for this. However, we have been keeping isort because ruff does not support Cython code but isort does. See this discussion for more info: [#14882 (comment)](https://github.com/rapidsai/cudf/issues/14882#issuecomment-1910887340)

@bdice the rest of the discussion in that thread does suggest that we'd be OK with making the switch and losing the limited Cython support. Maybe should go ahead with this now.

**Comment by bdice** [comment]
You’re right. I forgot the rest of that discussion. @vyasr If you’re in support, I am fine with it.

File: pyproject.toml
Comment (vyasr) [inline pyproject.toml:102]: Why is this rule ignored? Isn't using this the whole point of turning on the sorting functionality?

Comment (Borda) [reply pyproject.toml:102]: wanted to keep the diff minimal and just show the parity, so I may suggest landing this with this exception and then remove this line and apply related changes... or do you prefer to have it at once?

Comment (vyasr) [reply pyproject.toml:102]: I'm confused, what do you mean by parity? With this check disabled, won't we have strictly less linting done now than before? Unless I'm severely misunderstanding, disabling I001 disables the primary functionality of isort when running ruff: sorting imports in blocks.

Comment (Borda) [reply pyproject.toml:102]: so for now `isort` does not flag any issue so I set the Ruff so it does not either even it seem that Ruff found some issue...
yes, I agree, so I enable that rule and see how we like the formatting result and eventually close this PR if the formating is not liked :)

File: python/cudf/cudf/core/udf/masked_lowering.py
Comment (Borda) [inline python/cudf/cudf/core/udf/masked_lowering.py:1]: not sure ow this change happened but seem to be right...

**Comment by vyasr** [comment]
You'll notice that we have isort configuration in each of our Python package's pyproject.toml files, [like cudf's here](https://github.com/rapidsai/cudf/blob/branch-24.10/python/cudf/pyproject.toml#L82). We definitely want to preserve those rules, so you'll need to add those back as well. https://github.com/rapidsai/cudf/issues/14882#issuecomment-1910887340 discusses a bit how we could accomplish this in a way that shares most of the logic between the different packages in the repo. [ruff's docs also have some discussion of how to ensure near-equivalent results](https://docs.astral.sh/ruff/faq/#how-does-ruffs-import-sorting-compare-to-isort), and if you scroll down you'll see discussion of the first/third-party configuration and more. Hopefully that provides you a good enough starting point, but please follow up if you have more questions!

File: pyproject.toml
Comment (vyasr) [reply pyproject.toml]: My assumption was that without the check enabled ruff simply didn't find any issues, so no changes were made. It wouldn't be good if we merged the change and preserved the current sorting, only to find that nothing was getting sorted going forward :slightly_smiling_face:

**Comment by vyasr** [comment]
@Borda let us know if you need any help here. Please also retarget this PR at branch-24.12, we are on the verge of freezing 24.10.

**Comment by Borda** [comment]
> @Borda let us know if you need any help here. Please also retarget this PR at branch-24.12, we are on the verge of freezing 24.10.

sure will do it over weekend 🦩

**Comment by Borda** [comment]
> > Please also retarget this PR at branch-24.12, we are on the verge of freezing 24.10.
> 
> sure will do it over weekend 🦩

@vyasr rebased and updated (set I rule and run pre-commit on all files)

File: python/cudf/cudf/core/udf/masked_lowering.py
Comment (Borda) [reply python/cudf/cudf/core/udf/masked_lowering.py:1]: this is the result of running pre-commit on all files, so it updated files that have not been touched by any PR for a long time so pre-commit did not run

cc: @vyasr

**Comment by wence-** [comment]
Thanks, the next step here is to migrate the subproject-specific configuration for each python package, e.g. for cudf itself here https://github.com/rapidsai/cudf/blob/3420c71cb72f63db8d63164446cca042f354a08e/python/cudf/pyproject.toml#L84-L126 and replace it with the relevant project-specific `[tool.ruff.lint.isort]` configuration that mimics, as closely as possible, the settings. For example for the `cudf_polars` subproject, we have already done this, and you can see the classification here https://github.com/rapidsai/cudf/blob/3420c71cb72f63db8d63164446cca042f354a08e/python/cudf_polars/pyproject.toml#L172-L187

File: python/cudf/cudf/core/udf/masked_lowering.py
Comment (vyasr) [reply python/cudf/cudf/core/udf/masked_lowering.py:1]: It shouldn't trigger updates of files unless other parts of your git repo are out of date I believe. You may need to git fetch the remote or something. Files that you did not modify in this PR shouldn't get updated. If that's happening we can troubleshoot.

Comment (KyleFromNVIDIA) [reply python/cudf/cudf/core/udf/masked_lowering.py:1]: The copyright was updated because you changed another part of the file (added lines 10-11). This update should be correct.

**Comment by Borda** [comment]
@wence- just did and run applied to all files but seems that triggered some more formatting changes that were unintended 😕

File: python/cudf/pyproject.toml
Comment (wence-) [inline python/cudf/pyproject.toml:138]: I think we don't want this section to be populated (at least with these packages), because they are all handled by the `dask` and `rapids` sections you introduced below?

**Comment by wence-** [comment]
> @wence- just did and run applied to all files but seems that triggered some more formatting changes that were unintended 😕

Thanks. It's strange, because almost all the changes are _not_ in the import lists, but other formatting changes. So somehow the application of isort rules in ruff has changed what formatting it applied. Odd.

*question*: Did you run with `pre-commit run --all-files` or did you use a locally installed version of `ruff` to run the checks? If the latter, I think you would need to make sure the versions match up with what we specify in pre-commit.

File: python/cudf/pyproject.toml
Comment (wence-) [inline python/cudf/pyproject.toml:134]: For all these changes, we need to also add a section:

```suggestion
[tool.ruff]
extend = "../../pyproject.toml"

[tool.ruff.lint.isort]
```

So that we pick up the global configration as well.

**Comment by wence-** [comment]
> > @wence- just did and run applied to all files but seems that triggered some more formatting changes that were unintended 😕
> 
> Thanks. It's strange, because almost all the changes are _not_ in the import lists, but other formatting changes. So somehow the application of isort rules in ruff has changed what formatting it applied. Odd.
> 
> _question_: Did you run with `pre-commit run --all-files` or did you use a locally installed version of `ruff` to run the checks? If the latter, I think you would need to make sure the versions match up with what we specify in pre-commit.

Ah, I know what's going on. By introducing a ruff-specific section in the package-specific pyproject.toml files, we're now no longer picking up the "global" configuration (see https://docs.astral.sh/ruff/configuration/#config-file-discovery) for all the details.

tl;dr: the individual `pyproject.toml` files need to get an additional
```
[tool.ruff]
extend = "../../pyproject.toml"
```

**Comment by Borda** [comment]
> Did you run with `pre-commit run --all-files`

yes that is what I did

File: .github/workflows/pr.yaml
Comment (KyleFromNVIDIA) [inline .github/workflows/pr.yaml:48]: Reminder that this needs to be changed back before merging

**Comment by KyleFromNVIDIA** [comment]
@Borda Could you please resolve the merge conflicts?

**Comment by Borda** [comment]
> Could you please resolve the merge conflicts?

@KyleFromNVIDIA updated the pyproject.toml files and merged default branch

File: .github/workflows/pr.yaml
Comment (Borda) [reply .github/workflows/pr.yaml:48]:
```suggestion
    uses: rapidsai/shared-workflows/.github/workflows/changed-files.yaml@branch-24.12
```

Comment (KyleFromNVIDIA) [reply .github/workflows/pr.yaml]: This should not have been changed back yet. I was still in the process of testing it.

Comment (KyleFromNVIDIA) [reply .github/workflows/pr.yaml]: Test was successful. https://github.com/rapidsai/shared-workflows/pull/252 has been merged. I am reverting this change again.

File: python/dask_cudf/dask_cudf/backends.py
Comment (Borda) [inline python/dask_cudf/dask_cudf/backends.py:731]:
```suggestion
```

File: python/dask_cudf/dask_cudf/backends.py
Comment (Borda) [inline python/dask_cudf/dask_cudf/backends.py:737]:
```suggestion
            # CudfReadParquetPyarrowFS requires import of distributed beforehand
            # (See: https://github.com/dask/dask/issues/11352)
            import distributed  # noqa: F401
```

**Comment by bdice** [comment]
@Borda I'm going to trigger tests to run -- we can hold off on merging the upstream for now, unless there are conflicts.

File: docs/cudf/source/developer_guide/contributing_guide.md
Comment (bdice) [inline docs/cudf/source/developer_guide/contributing_guide.md:18]: We can update this to a canonical URL now.
```suggestion
- [`ruff`](https://docs.astral.sh/ruff/) checks for general code formatting compliance.
```

**Review by bdice — COMMENTED** [review]
I have a few local changes to push, then will re-trigger CI. I think this is nearly ready?

**Review by bdice — APPROVED** [review]
I'm happy with this, assuming that CI passes. @vyasr @wence- Can you give a final check, so we don't hold this PR in limbo? (it has been requiring a lot of upstream-merges and conflict-resolution)

File: python/cudf/pyproject.toml
Comment (vyasr) [inline python/cudf/pyproject.toml:139]: Is there a reason that we set this to true? From a quick look at [the docs](https://docs.astral.sh/ruff/settings/#lint_isort_force-sort-within-sections) I suspect that this configuration is responsible for the vast majority of the changes in this PR. I'm not opposed to it, just wondering if we discussed it?

Comment (bdice) [reply python/cudf/pyproject.toml:139]: This does seem to be a change from the status quo -- I overlooked this earlier, and wondered if it was just a change in ruff relative to isort. I will remove this option but we can wait for @wence- to confirm if he prefers something else before merging.

Comment (bdice) [reply python/cudf/pyproject.toml]: Before: +554 −610
After: +318 −380

This reduces the diff a lot, which is good.

Comment (wence-) [reply python/cudf/pyproject.toml]: I had used it in cudf-polars but I have no string opinions

Comment (Borda) [reply python/cudf/pyproject.toml]: seems I misinterpreted this options with sort, thx for fixing it

**Review by vyasr — APPROVED** [review]
Cool this LGTM now.

**Review by jakirkham — APPROVED** [review]

**Comment by wence-** [comment]
Thanks @Borda!

**Comment by vyasr** [comment]
Thanks for this @Borda!

**Pull request merged**
